### PR TITLE
[pytorch] Migrate away from using legacy Variable constructor in test_nn.py

### PR DESCRIPTION
**Pull request by rohan-varma**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#26077 [pytorch] Migrate away from using Variable( in test_nn.py**

As per #26071, we would like to get rid of the calls to Variable(
where possible. This diff removes the calls in the test file test_nn.py. The
unit tests should all still pass as expected.

Differential Revision: [D17336484](https://our.internmc.facebook.com/intern/diff/D17336484/)

### CODE REVIEW

**Review by soumith — APPROVED** [review]

**Review by soumith — CHANGES_REQUESTED** [review]
Variable shouldn't be replaced by torch.Tensor which itself is a legacy constructor. You should pass `requires_grad` directly into constructors like `torch.empty`, `torch.randn`, etc.

File: test/test_nn.py [hunk 1]
```diff
@@ -3453,7 +3453,7 @@ def _test_GroupNorm_general(self, device="cpu", dtype=torch.float):
             self.assertRaises(RuntimeError, lambda: gn(input))
 
     def _test_GroupNorm_cuda_half(self):
-        input = Variable(torch.empty(2, 3, 3, 2).to("cuda", torch.half).random_(1, 10), requires_grad=True)
+        input = torch.empty(2, 3, 3, 2, requires_grad=True).to("cuda", torch.half).random_(1, 10)
```
Comment (rohan-varma) [inline test/test_nn.py]: Could we just remove this line? `input` is already being overwritten by the very next line.

Comment (zou3519) [reply test/test_nn.py]: Go for it. It doesn't look significant.

File: test/test_nn.py [hunk 2]
```diff
@@ -2353,7 +2353,7 @@ def test_threshold_int(self):
 
     def test_embedding_sparse_basic(self):
         embedding = nn.Embedding(10, 20, sparse=True)
-        input = Variable(torch.LongTensor([[0, 2, 4, 5], [4, 3, 0, 9]]))
+        input = torch.LongTensor([[0, 2, 4, 5], [4, 3, 0, 9]])
```
Comment (zou3519) [inline test/test_nn.py]: we should try to use new-style constructors. All of the `torch.{Type}Tensor` constructors are deprecated. The replacement for `LongTensor` is `torch.tensor(data, dtype=torch.long)`

File: test/test_nn.py [hunk 3]
```diff
@@ -2449,27 +2449,27 @@ def _test_embedding_backward(self, device='cpu', dtype=torch.float64):
 
     def test_embedding_padding_idx(self):
         embedding = nn.Embedding(10, 20, padding_idx=0)
-        input = Variable(torch.LongTensor([[0, 2, 4, 5], [4, 3, 0, 9]]))
+        input = torch.LongTensor([[0, 2, 4, 5], [4, 3, 0, 9]])
```
Comment (zou3519) [inline test/test_nn.py]: ditto

File: test/test_nn.py [hunk 4]
```diff
@@ -3390,7 +3390,7 @@ def _test_LayerNorm_general(self, device="cpu", dtype=torch.float):
             self.assertRaises(RuntimeError, lambda: ln(input))
 
     def _test_LayerNorm_cuda_half(self):
-        input = Variable(torch.empty(2, 3, 3, 2).to("cuda", torch.half).random_(1, 10), requires_grad=True)
+        input = torch.empty(2, 3, 3, 2).to("cuda", torch.half).random_(1, 10).requires_grad_(True)
```
Comment (zou3519) [inline test/test_nn.py]: We can combine `empty` and `to` in the following manner:
`torch.empty(2, 3, 3, 2, device='cuda', dtype=torch.half)`. It's more efficient because we don't incur the host-to-device transfer.

File: test/test_nn.py [hunk 5]
```diff
@@ -4262,7 +4261,7 @@ def test_broadcast_no_grad(self):
     @unittest.skipIf(not TEST_MULTIGPU, "multi-GPU not supported")
     def test_replicate(self):
         module = nn.Linear(10, 5).float().cuda()
-        input = Variable(torch.randn(2, 10).float().cuda())
+        input = torch.randn(2, 10).float().cuda()
```
Comment (zou3519) [inline test/test_nn.py]: `torch.randn(2, 10, dtype=torch.float, device='cuda')`

File: test/test_nn.py [hunk 6]
```diff
@@ -9055,9 +9054,9 @@ def test_conv_double_backward_stride(self):
     def test_cudnn_noncontiguous_weight(self):
         # Noncontiguous weights must be contiguous() before being
         # passed to cuDNN
-        input = Variable(torch.cuda.DoubleTensor([1, 1, 1]).view(1, 1, 3))
-        weights1 = Variable(torch.cuda.DoubleTensor([1]).expand(1, 1, 2))
-        weights2 = Variable(torch.cuda.DoubleTensor([1]).expand(1, 1, 2)).contiguous()
+        input = torch.cuda.DoubleTensor([1, 1, 1]).view(1, 1, 3)
```
Comment (zou3519) [inline test/test_nn.py]: Replace torch.cuda.DoubleTensor with `torch.tensor(data, device='cuda', dtype=torch.double)`

**Review by zou3519 — CHANGES_REQUESTED** [review]
Thank you for doing this, @rohan-varma. All of the change are correct, but it would be great if we could clean up the style a bit. I had two main comments:

1) In doing this porting we should aim to use as few legacy constructors as possible. Those can be identified by how they have "Tensor" (with a capital T) in the name. i.e., `LongTensor`, `DoubleTensor`.

2) The new-style tensor constructors (`tensor`, `randn`, etc) can take the device type and dtype as so: `torch.tensor(data, device=device, dtype=dtype)`. We should try to avoid `torch.tensor(...).to(...)` calls and instead fuse them into a single `torch.tensor(...)` call. This is more efficient because we can end up creating the data directly on the device most of the time.

**Comment by rohan-varma** [comment]
> 1. In doing this porting we should aim to use as few legacy constructors as possible. Those can be identified by how they have "Tensor" (with a capital T) in the name. i.e., `LongTensor`, `DoubleTensor`.

Sounds good. I think for this PR, I'll focus on the legacy constructors that are used in the changed lines, since this is a large test file and there are many uses of the legacy constructors (without `Variable`). For example, there are 69 uses of `torch.Tensor`. To keep the PRs relatively small and avoid merge conflicts, I'll send a follow up PR fixing those.

File: test/test_nn.py [hunk 7]
```diff
@@ -3271,7 +3271,7 @@ def _test_InstanceNorm_general(self, cls, input, device="cpu", dtype=torch.float
 
     def _test_InstanceNorm_cuda_half(self, cls, input):
         # THNN
-        input = Variable(input.cuda().half().random_(1, 10), requires_grad=True)
+        input = input.cuda().half().random_(1, 10).requires_grad_(True)
```
Comment (zou3519) [inline test/test_nn.py]: You missed this line

**Review by zou3519 — APPROVED** [review]

File: test/test_nn.py (hunk 7 above)
Comment (rohan-varma) [reply test/test_nn.py]: It looks like this line is overwriting the input tensor called `input`. Not sure how to change this as we're not explicitly constructing a tensor here, would we want to do something like `input = torch.tensor(input.clone().detach(), device="cuda", dtype=torch.half).random_(1,10).requires_grad_(True)`?

Comment (zou3519) [reply test/test_nn.py]: Let's do `input.to(device='cuda', dtype='half').random_(...)...` to get it all done in one go.
